### DISCUSSION TITLE
exploring prior art of merging techniques

### DISCUSSION THREAD

**Discussion by joyously** [jj-vcs/jj#3202] [General]
I've read in the [Git manual](https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging) that it doesn't make guesses when merging. If it's not obvious, it is a conflict.
And yet, there are (perhaps old) stories of Git messing up: 
[When a Clean Merge is Wrong](https://www.caktusgroup.com/blog/2018/03/19/when-clean-merge-wrong/)
[Is it possible for Git merging to make a mistake without detecting a conflict?](https://www.quora.com/Is-it-possible-for-Git-merging-to-make-a-mistake-without-detecting-a-conflict)

So I assume there is some magic code in `jj` to be able to do all the rebases and rewriting that it does. Wasn't there a fairly recent comment about maybe needing more information for the more complex cases? I know Darcs and Pijul can compute the opposite of a change and that helps somehow (I'm not familiar with the algorithm even).

I was reading about [AutoMerge](https://automerge.org/), which uses CRDT for collaborative editing, and their docs state the[ rules it uses for its basic types to merge](https://automerge.org/docs/under-the-hood/merge_rules/). I wonder if this is rigorous enough for code.

**Comment by amiryal** [comment]
With regards to the question:

> Wasn't there a fairly recent comment about maybe needing more information for the more complex cases?

I did mention it in passing in https://github.com/martinvonz/jj/pull/3109#discussion_r1498875908 (look for “side note” and “pathological cases”). That comment includes a couple more links to such pathological cases with extensive commentary:
https://tahoe-lafs.org/~zooko/badmerge/simple.html
https://tahoe-lafs.org/~zooko/badmerge/concrete-good-semantics.html

**Comment by amiryal** [comment]
> [When a Clean Merge is Wrong](https://www.caktusgroup.com/blog/2018/03/19/when-clean-merge-wrong/)

Thanks for sharing this one!
